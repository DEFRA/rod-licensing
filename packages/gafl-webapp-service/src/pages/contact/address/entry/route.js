import { ADDRESS_ENTRY, ADDRESS_LOOKUP } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { countries } from '../../../../processors/refdata-helper.js'
import { nextPage } from '../../../../routes/next-page.js'
import { getPronoun } from '../../../../processors/licence-type-display.js'

export const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  street: validation.contact.createStreetValidator(Joi),
  locality: validation.contact.createLocalityValidator(Joi),
  town: validation.contact.createTownValidator(Joi),
  postcode: Joi.alternatives().conditional('country-code', {
    is: Joi.string().pattern(/^GB/),
    then: validation.contact.createUKPostcodeValidator(Joi),
    otherwise: validation.contact.createOverseasPostcodeValidator(Joi)
  }),
  'country-code': Joi.string().required()
}).options({ abortEarly: false, allowUnknown: true })

export const getCountryDropDownOptions = async () => {
  const options = await countries.getAll()
  return options.filter(country => country.code !== 'GB')
}

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  const pronoun = getPronoun(isLicenceForYou).possessive

  return {
    pronoun,
    searchTerms: !addresses?.length && searchTerms ? searchTerms : null,
    countries: await getCountryDropDownOptions(),
    lookupPage: ADDRESS_LOOKUP.uri
  }
}

export default pageRoute(ADDRESS_ENTRY.page, ADDRESS_ENTRY.uri, validator, nextPage, getData)
