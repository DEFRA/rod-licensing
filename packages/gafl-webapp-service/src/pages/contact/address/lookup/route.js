import { ADDRESS_LOOKUP, ADDRESS_ENTRY, OS_TERMS } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { hasJunior, validation } from '@defra-fish/business-rules-lib'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  postcode: validation.contact.createUKPostcodeValidator(Joi)
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou: permission.isLicenceForYou,
    licenceLength: permission.licenceLength,
    junior: hasJunior(permission),
    isPhysical: isPhysical(permission),
    uri: {
      entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri),
      osTerms: addLanguageCodeToUri(request, OS_TERMS.uri)
    }
  }
}

export default pageRoute(ADDRESS_LOOKUP.page, ADDRESS_LOOKUP.uri, validator, nextPage, getData)
