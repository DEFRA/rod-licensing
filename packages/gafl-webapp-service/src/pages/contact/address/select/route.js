import { ADDRESS_SELECT, ADDRESS_LOOKUP, ADDRESS_ENTRY } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou,
    addresses,
    searchTerms,
    lookupPage: addLanguageCodeToUri(request, ADDRESS_LOOKUP.uri),
    entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri)
  }
}

export default pageRoute(
  ADDRESS_SELECT.page,
  ADDRESS_SELECT.uri,
  Joi.object({
    address: Joi.number().integer().min(0).required()
  }).options({ abortEarly: false, allowUnknown: true }),
  nextPage,
  getData
)
