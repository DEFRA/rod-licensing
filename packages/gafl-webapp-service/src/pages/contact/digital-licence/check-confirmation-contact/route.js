import { CHECK_CONFIRMATION_CONTACT, CONTACT, LICENCE_CONFIRMATION_METHOD } from '../../../../uri.js'
import { HOW_CONTACTED } from '../../../../processors/mapping-constants.js'
import GetDataRedirect from '../../../../handlers/get-data-redirect.js'
import pageRoute from '../../../../routes/page-route.js'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

export const getData = async request => {
  const { licensee } = await request.cache().helpers.transaction.getCurrentPermission()

  if (licensee.preferredMethodOfConfirmation === HOW_CONTACTED.none) {
    throw new GetDataRedirect(CONTACT.uri)
  }

  const whatToChange = licensee.preferredMethodOfConfirmation === HOW_CONTACTED.email ? '?change=email' : '?change=mobile'
  const change = addLanguageCodeToUri(request, `${LICENCE_CONFIRMATION_METHOD.uri}${whatToChange}`)

  return {
    licensee,
    uri: {
      contact: addLanguageCodeToUri(request, CONTACT.uri),
      change
    }
  }
}

export default pageRoute(CHECK_CONFIRMATION_CONTACT.page, CHECK_CONFIRMATION_CONTACT.uri, null, nextPage, getData)
