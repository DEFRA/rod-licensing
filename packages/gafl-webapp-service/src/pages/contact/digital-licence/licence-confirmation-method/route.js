import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'

import { CONTACT, LICENCE_CONFIRMATION_METHOD } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import GetDataRedirect from '../../../../handlers/get-data-redirect.js'
import { nextPage } from '../../../../routes/next-page.js'
import { HOW_CONTACTED } from '../../../../processors/mapping-constants.js'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import { mobilePhoneValidator } from '../../../../processors/contact-validator.js'

const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  // page is only permitted for physical licences
  if (!isPhysical(permission)) {
    throw new GetDataRedirect(CONTACT.uri)
  }

  return {
    licensee: permission.licensee,
    howContacted: HOW_CONTACTED
  }
}

const validator = Joi.object({
  'licence-confirmation-method': Joi.string()
    .valid('email', 'text', 'none')
    .required(),
  text: Joi.alternatives().conditional('licence-confirmation-method', {
    is: 'text',
    then: mobilePhoneValidator,
    otherwise: Joi.string().empty('')
  }),
  email: Joi.alternatives().conditional('licence-confirmation-method', {
    is: 'email',
    then: validation.contact.createEmailValidator(Joi),
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(LICENCE_CONFIRMATION_METHOD.page, LICENCE_CONFIRMATION_METHOD.uri, validator, nextPage, getData)
