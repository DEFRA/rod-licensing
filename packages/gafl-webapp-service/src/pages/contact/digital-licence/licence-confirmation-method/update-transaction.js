import { LICENCE_CONFIRMATION_METHOD } from '../../../../uri.js'
import { HOW_CONTACTED } from '../../../../processors/mapping-constants.js'

export default async request => {
  const { payload } = await request.cache().helpers.page.getCurrentPermission(LICENCE_CONFIRMATION_METHOD.page)
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const { licensee } = permission

  switch (payload['licence-confirmation-method']) {
    case 'email':
      licensee.preferredMethodOfConfirmation = HOW_CONTACTED.email
      licensee.email = payload.email
      break
    case 'text':
      licensee.preferredMethodOfConfirmation = HOW_CONTACTED.text
      licensee.mobilePhone = payload.text
      break
    default:
      licensee.preferredMethodOfConfirmation = HOW_CONTACTED.none
      break
  }

  await request.cache().helpers.transaction.setCurrentPermission(permission)
}
