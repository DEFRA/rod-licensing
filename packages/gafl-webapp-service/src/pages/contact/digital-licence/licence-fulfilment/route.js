import Joi from 'joi'

import pageRoute from '../../../../routes/page-route.js'
import { CONTACT, LICENCE_FULFILMENT } from '../../../../uri.js'
import { nextPage } from '../../../../routes/next-page.js'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import GetDataRedirect from '../../../../handlers/get-data-redirect.js'

const isPhysicalPermission = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  // page is only permitted for physical licences
  if (!isPhysical(permission)) {
    throw new GetDataRedirect(CONTACT.uri)
  }
}

const validator = Joi.object({
  'licence-option': Joi.string()
    .valid('digital', 'paper-licence')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(LICENCE_FULFILMENT.page, LICENCE_FULFILMENT.uri, validator, nextPage, isPhysicalPermission)
