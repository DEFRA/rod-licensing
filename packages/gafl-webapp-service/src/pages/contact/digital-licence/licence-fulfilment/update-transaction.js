import { LICENCE_FULFILMENT } from '../../../../uri.js'

export default async request => {
  const { payload } = await request.cache().helpers.page.getCurrentPermission(LICENCE_FULFILMENT.page)
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const { licensee } = permission

  const licenceOption = payload['licence-option']
  if (licenceOption === 'digital') {
    licensee.postalFulfilment = false
  }
  if (licenceOption === 'paper-licence') {
    licensee.postalFulfilment = true
  }

  await request.cache().helpers.transaction.setCurrentPermission(permission)
}
