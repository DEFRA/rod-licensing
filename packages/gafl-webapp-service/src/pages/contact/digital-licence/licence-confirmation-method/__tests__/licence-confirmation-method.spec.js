import {
  LICENCE_LENGTH,
  DATE_OF_BIRTH,
  ADDRESS_ENTRY,
  LICENCE_CONFIRMATION_METHOD,
  CONTACT,
  CONTACT_SUMMARY,
  LICENCE_SUMMARY,
  LICENCE_TO_START,
  LICENCE_TYPE,
  NAME,
  NEWSLETTER,
  NEW_TRANSACTION,
  TEST_TRANSACTION
} from '../../../../../uri.js'

import { HOW_CONTACTED } from '../../../../../processors/mapping-constants.js'
import { start, stop, initialize, injectWithCookies } from '../../../../../__mocks__/test-utils-system.js'
import { ADULT_TODAY, dobHelper } from '../../../../../__mocks__/test-utils-business-rules'
import { licenceToStart } from '../../../../licence-details/licence-to-start/update-transaction'
import { licenseTypes } from '../../../../licence-details/licence-type/route'

beforeAll(d => start(d))
beforeAll(d => initialize(d))
afterAll(d => stop(d))

const goodAddress = {
  premises: '14 HOWECROFT COURT',
  street: 'EASTMEAD LANE',
  locality: '',
  town: 'BRISTOL',
  postcode: 'BS9 1HJ',
  'country-code': 'GB'
}

describe('The licence confirmation method page', () => {
  describe('for a full 12 month licence, adult', () => {
    beforeAll(async d => {
      await injectWithCookies('GET', NEW_TRANSACTION.uri)
      await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(ADULT_TODAY))
      await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
      await injectWithCookies('POST', LICENCE_LENGTH.uri, { 'licence-length': '12M' })
      d()
    })

    it('return the page on request', async () => {
      const response = await injectWithCookies('GET', LICENCE_CONFIRMATION_METHOD.uri)
      expect(response.statusCode).toBe(200)
    })

    it('redirects to itself on an empty response', async () => {
      const response = await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, {})
      expect(response.statusCode).toBe(302)
      expect(response.headers.location).toBe(LICENCE_CONFIRMATION_METHOD.uri)
    })

    it('redirects to itself on an invalid contact method', async () => {
      const response = await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, { 'licence-confirmation-method': 'skype' })
      expect(response.statusCode).toBe(302)
      expect(response.headers.location).toBe(LICENCE_CONFIRMATION_METHOD.uri)
    })

    it('redirects to itself on an empty email', async () => {
      const response = await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, {
        'licence-confirmation-method': 'email',
        email: ''
      })
      expect(response.statusCode).toBe(302)
      expect(response.headers.location).toBe(LICENCE_CONFIRMATION_METHOD.uri)
    })

    it('redirects to itself on an invalid email', async () => {
      const response = await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, {
        'licence-confirmation-method': 'email',
        email: 'foo'
      })
      expect(response.statusCode).toBe(302)
      expect(response.headers.location).toBe(LICENCE_CONFIRMATION_METHOD.uri)
    })

    it('redirects to itself on an empty mobile number', async () => {
      const response = await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, { 'licence-confirmation-method': 'text', text: '' })
      expect(response.statusCode).toBe(302)
      expect(response.headers.location).toBe(LICENCE_CONFIRMATION_METHOD.uri)
    })

    it.each(['+44(0)7513438168', '923246734', 'email@com', '01179835413', '+457513 438 167'])(
      'redirects to itself on an invalid mobile number: %s',
      async mobileNumber => {
        const response = await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, {
          'licence-confirmation-method': 'text',
          text: mobileNumber
        })
        expect(response.statusCode).toBe(302)
        expect(response.headers.location).toBe(LICENCE_CONFIRMATION_METHOD.uri)
      }
    )

    it('post response email sets licence-confirmation-method - email, in the cache', async () => {
      await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, {
        'licence-confirmation-method': 'email',
        email: 'example@email.com'
      })
      const { payload } = await injectWithCookies('GET', TEST_TRANSACTION.uri)
      expect(JSON.parse(payload).permissions[0].licensee.preferredMethodOfConfirmation).toEqual(HOW_CONTACTED.email)
      expect(JSON.parse(payload).permissions[0].licensee.email).toEqual('example@email.com')
    })

    it.each([
      ['07513438122', '07513438122'],
      ['07513 438167', '07513438167'],
      ['07513 438 167', '07513438167'],
      ['+447513 438 167', '+447513438167'],
      ['+44 7513 438 167', '+447513438167'],
      ['07513438168', '07513438168']
    ])(
      'post response text sets licence-confirmation-method - text in the cache for mobile number %s',
      async (mobileNumberi, mobileNumbero) => {
        await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, { 'licence-confirmation-method': 'text', text: mobileNumberi })
        const { payload } = await injectWithCookies('GET', TEST_TRANSACTION.uri)
        expect(JSON.parse(payload).permissions[0].licensee.preferredMethodOfConfirmation).toEqual(HOW_CONTACTED.text)
        expect(JSON.parse(payload).permissions[0].licensee.mobilePhone).toEqual(mobileNumbero)
      }
    )
  })

  describe('for 1 day licence', () => {
    beforeAll(async d => {
      await injectWithCookies('GET', NEW_TRANSACTION.uri)
      await injectWithCookies('POST', DATE_OF_BIRTH.uri, dobHelper(ADULT_TODAY))
      await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
      await injectWithCookies('POST', LICENCE_LENGTH.uri, { 'licence-length': '1D' })
      d()
    })

    it('redirects to the contact page', async () => {
      const response = await injectWithCookies('GET', LICENCE_CONFIRMATION_METHOD.uri)
      expect(response.statusCode).toBe(302)
      expect(response.headers.location).toBe(CONTACT.uri)
    })
  })

  describe('if the contact summary has been seen', () => {
    it('controller redirects to the contact page', async () => {
      // Set up the licence details
      await injectWithCookies('POST', LICENCE_TO_START.uri, { 'licence-to-start': licenceToStart.AFTER_PAYMENT })
      await injectWithCookies('POST', LICENCE_TYPE.uri, { 'licence-type': licenseTypes.troutAndCoarse2Rod })
      await injectWithCookies('POST', LICENCE_SUMMARY.uri)

      // Set up the contact details
      await injectWithCookies('POST', NAME.uri, { 'last-name': 'Graham', 'first-name': 'Willis' })
      await injectWithCookies('POST', ADDRESS_ENTRY.uri, goodAddress)
      await injectWithCookies('POST', CONTACT.uri, { 'how-contacted': 'email', email: 'new3@example.com' })
      await injectWithCookies('POST', NEWSLETTER.uri, { newsletter: 'yes', 'email-entry': 'no' })

      await injectWithCookies('GET', CONTACT_SUMMARY.uri)
      const response = await injectWithCookies('POST', LICENCE_CONFIRMATION_METHOD.uri, {
        'licence-confirmation-method': 'email',
        email: 'example@email.com'
      })
      expect(response.statusCode).toBe(302)
      expect(response.headers.location).toBe(CONTACT.uri)
    })
  })
})
