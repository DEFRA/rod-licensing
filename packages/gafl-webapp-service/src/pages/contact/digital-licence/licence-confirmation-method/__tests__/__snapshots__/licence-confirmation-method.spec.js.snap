// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.getData returns the expected data 1`] = `
Object {
  "howContacted": Object {
    "email": "Email",
    "letter": "Letter",
    "none": "Prefer not to be contacted",
    "text": "Text",
  },
  "licensee": Object {
    "firstName": "Lando",
    "lastName": "Norris",
  },
}
`;
